File: native/android/compose/src/main/kotlin/dev/respondkit/compose/RespondKitScreen.kt
package dev.respondkit.compose

import androidx.activity.compose.BackHandler
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.itemsIndexed
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.text.selection.SelectionContainer
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.focus.onFocusChanged
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.graphics.compositeOver
import androidx.compose.ui.graphics.luminance
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleEventObserver
import androidx.lifecycle.compose.LocalLifecycleOwner
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import dev.respondkit.core.*
import java.time.format.DateTimeFormatter
import java.time.format.FormatStyle
import java.util.Locale
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.launch

private val WidgetInk = Color(0xFF171717)
private val WidgetMuted = Color(0xFF737373)
private val WidgetBorder = Color(0xFFE5E5E5)
private val WidgetFill = Color(0xFFF5F5F5)
private val WidgetError = Color(0xFFDC2626)

/** Mount once at the app root, not inside the conditionally presented support destination. */
@Composable
fun RespondKitLifecycle(store: RespondKitStore) {
    val lifecycle = LocalLifecycleOwner.current.lifecycle
    DisposableEffect(store, lifecycle) {
        val observer = LifecycleEventObserver { _, _ ->
            store.setForeground(lifecycle.currentState.isAtLeast(Lifecycle.State.RESUMED))
        }
        lifecycle.addObserver(observer)
        store.setForeground(lifecycle.currentState.isAtLeast(Lifecycle.State.RESUMED))
        onDispose {
            lifecycle.removeObserver(observer)
            store.setForeground(false)
        }
    }
}

/**
 * One full-screen conversation. Accent inherits the host; neutral surfaces match the web widget.
 */
@Composable
fun RespondKitScreen(
    store: RespondKitStore,
    onClose: () -> Unit,
    modifier: Modifier = Modifier,
    title: String = stringResource(R.string.respondkit_support),
    accentColor: Color? = null,
) {
    val accent = (accentColor ?: MaterialTheme.colorScheme.primary).compositeOver(Color.White)
    MaterialTheme(
        colorScheme =
            lightColorScheme(
                primary = accent,
                onPrimary = if (accent.luminance() > 0.179f) Color.Black else Color.White,
                primaryContainer = accent.copy(alpha = 0.1f).compositeOver(Color.White),
                onPrimaryContainer = WidgetInk,
                background = Color.White,
                surface = Color.White,
                onSurface = WidgetInk,
                surfaceVariant = WidgetFill,
                onSurfaceVariant = WidgetMuted,
                outline = WidgetBorder,
                error = WidgetError,
            )
    ) {
        RespondKitContent(store, onClose, modifier, title)
    }
}

@Composable
private fun RespondKitContent(
    store: RespondKitStore,
    onClose: () -> Unit,
    modifier: Modifier,
    title: String,
) {
    val state by store.state.collectAsStateWithLifecycle()
    val scope = rememberCoroutineScope()
    DisposableEffect(store) {
        store.setScreenVisible(true)
        onDispose { store.setScreenVisible(false) }
    }
    LaunchedEffect(store) { store.openConversation() }
    BackHandler(onBack = onClose)
    Column(modifier.fillMaxSize().background(Color.White).safeDrawingPadding().imePadding()) {
        Row(
            Modifier.fillMaxWidth()
                .heightIn(min = 64.dp)
                .padding(horizontal = 16.dp, vertical = 12.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Column(Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(0.dp)) {
                Text(
                    title,
                    fontSize = 16.sp,
                    fontWeight = FontWeight.SemiBold,
                    color = WidgetInk,
                    maxLines = 1,
                )
                Text("Ask us anything", fontSize = 14.sp, lineHeight = 20.sp, color = WidgetMuted)
            }
            IconButton(
                onClick = onClose,
                modifier = Modifier.size(40.dp).testTag("respondkit-close"),
            ) {
                Icon(
                    painterResource(R.drawable.respondkit_close),
                    "Close support chat",
                    Modifier.size(16.dp),
                    tint = WidgetInk,
                )
            }
        }
        HorizontalDivider(color = WidgetBorder)
        state.errorMessage?.let { error ->
            Row(
                Modifier.fillMaxWidth().background(WidgetFill).padding(12.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Text(error, Modifier.weight(1f), fontSize = 14.sp, color = WidgetInk)
                TextButton(onClick = { scope.launch { store.refresh() } }) { Text("Retry") }
            }
        }
        Conversation(store, state, Modifier.weight(1f))
    }
}

@Composable
private fun Conversation(store: RespondKitStore, state: SupportState, modifier: Modifier) {
    val scope = rememberCoroutineScope()
    val listState = rememberLazyListState()
    val rows = remember(state.messages, state.pendingMessages) { transcriptRows(state) }
    val bottomKey = "bottom-${state.loadedCursor}"
    var atBottom by remember(state.activeThreadId) { mutableStateOf(true) }
    var unseen by remember(state.activeThreadId) { mutableIntStateOf(0) }
    var previousIds by remember(state.activeThreadId) { mutableStateOf(emptyList<String>()) }
    val ids = rows.map { it.id }
    LaunchedEffect(state.activeThreadId, ids) {
        if (ids.isNotEmpty()) {
            if (previousIds.isEmpty() || atBottom) {
                listState.scrollToItem(rows.size)
                unseen = 0
            } else {
                unseen += ids.count { it !in previousIds }
            }
        }
        previousIds = ids
    }
    LaunchedEffect(state.activeThreadId, state.loadedCursor, state.isForeground, ids.isEmpty()) {
        snapshotFlow {
                listState.layoutInfo.let { layout ->
                    layout.visibleItemsInfo.any {
                        it.key == bottomKey &&
                            it.offset >= layout.viewportStartOffset &&
                            it.offset + it.size <= layout.viewportEndOffset
                    }
                }
            }
            .distinctUntilChanged()
            .collect { visible ->
                atBottom = visible || ids.isEmpty()
                if (visible) {
                    unseen = 0
                    state.activeThreadId?.let { store.markDisplayed(it, state.loadedCursor) }
                }
            }
    }
    Column(modifier.fillMaxWidth()) {
        Box(Modifier.weight(1f).fillMaxWidth()) {
            LazyColumn(
                Modifier.fillMaxSize().testTag("respondkit-messages"),
                state = listState,
                contentPadding = PaddingValues(horizontal = 16.dp, vertical = 20.dp),
                verticalArrangement = Arrangement.spacedBy(12.dp, Alignment.Bottom),
            ) {
                itemsIndexed(rows, key = { _, row -> row.id }) { index, row ->
                    Column {
                        if (
                            row.date != null &&
                                (index == 0 ||
                                    rows[index - 1].date?.toLocalDate() != row.date.toLocalDate())
                        ) {
                            Text(
                                row.date.format(
                                    DateTimeFormatter.ofLocalizedDate(FormatStyle.MEDIUM)
                                        .withLocale(Locale.getDefault())
                                ),
                                Modifier.fillMaxWidth().padding(vertical = 8.dp),
                                textAlign = TextAlign.Center,
                                fontSize = 12.sp,
                                lineHeight = 16.sp,
                                color = WidgetMuted,
                            )
                            Spacer(Modifier.height(12.dp))
                        }
                        Bubble(row, state.isSending) { scope.launch { store.retry(it) } }
                    }
                }
                item(key = bottomKey) { Spacer(Modifier.height(2.dp)) }
            }
            if (rows.isEmpty()) {
                if (state.isLoading && !state.isSending) {
                    Column(
                        Modifier.fillMaxWidth().align(Alignment.TopStart).padding(16.dp).semantics {
                            contentDescription = "Loading messages"
                        },
                        verticalArrangement = Arrangement.spacedBy(16.dp),
                    ) {
                        listOf(0.75f to 56.dp, 0.8f to 80.dp, 0.67f to 48.dp).forEachIndexed {
                            index,
                            (width, height) ->
                            Box(
                                Modifier.fillMaxWidth(width)
                                    .height(height)
                                    .align(if (index == 1) Alignment.Start else Alignment.End)
                                    .background(WidgetFill, RoundedCornerShape(10.dp))
                            )
                        }
                    }
                } else
                    Column(
                        Modifier.align(Alignment.Center).padding(horizontal = 16.dp),
                        horizontalAlignment = Alignment.CenterHorizontally,
                        verticalArrangement = Arrangement.spacedBy(4.dp),
                    ) {
                        Text(
                            "How can we help?",
                            fontSize = 14.sp,
                            fontWeight = FontWeight.Medium,
                            color = WidgetInk,
                        )
                        Text(
                            "Send a message and keep this page open for a quick reply.",
                            fontSize = 14.sp,
                            color = WidgetMuted,
                            textAlign = TextAlign.Center,
                        )
                    }
            }
            if (unseen > 0)
                OutlinedButton(
                    onClick = {
                        scope.launch {
                            listState.animateScrollToItem(rows.size)
                            unseen = 0
                        }
                    },
                    modifier =
                        Modifier.align(Alignment.BottomEnd).padding(12.dp).semantics {
                            contentDescription = "Latest messages"
                        },
                    colors =
                        ButtonDefaults.outlinedButtonColors(
                            containerColor = Color.White,
                            contentColor = WidgetInk,
                        ),
                ) {
                    Icon(painterResource(R.drawable.respondkit_down), null, Modifier.size(16.dp))
                    Spacer(Modifier.width(6.dp))
                    Text("$unseen new", fontSize = 14.sp)
                }
        }
        HorizontalDivider(color = WidgetBorder)
        if (state.activeThread?.state == "closed") {
            Column(Modifier.padding(12.dp)) {
                Text(
                    stringResource(R.string.respondkit_closed_help),
                    fontSize = 14.sp,
                    color = WidgetMuted,
                )
                TextButton(onClick = { store.selectThread(null) }) {
                    Text(stringResource(R.string.respondkit_another_message))
                }
            }
        } else Composer(store, state)
    }
}

@Composable
private fun Composer(store: RespondKitStore, state: SupportState) {
    val scope = rememberCoroutineScope()
    var focused by remember { mutableStateOf(false) }
    val canSend =
        !state.isLoading &&
            !state.isSending &&
            state.draft.isNotBlank() &&
            state.draft.length <= 6_000
    Row(
        Modifier.fillMaxWidth().padding(12.dp),
        horizontalArrangement = Arrangement.spacedBy(8.dp),
        verticalAlignment = Alignment.Bottom,
    ) {
        BasicTextField(
            state.draft,
            store::setDraft,
            Modifier.weight(1f)
                .heightIn(min = 44.dp)
                .testTag("respondkit-composer")
                .semantics { contentDescription = "Message" }
                .onFocusChanged { focused = it.isFocused }
                .border(
                    1.dp,
                    if (focused) MaterialTheme.colorScheme.primary else WidgetBorder,
                    RoundedCornerShape(12.dp),
                )
                .padding(horizontal = 10.dp, vertical = 10.dp),
            textStyle = TextStyle(fontSize = 16.sp, lineHeight = 20.sp, color = WidgetInk),
            cursorBrush = SolidColor(MaterialTheme.colorScheme.primary),
            maxLines = 6,
            decorationBox = { input ->
                Box {
                    if (state.draft.isEmpty())
                        Text(
                            "Write a message…",
                            fontSize = 16.sp,
                            lineHeight = 20.sp,
                            color = WidgetMuted,
                        )
                    input()
                }
            },
        )
        Button(
            onClick = { scope.launch { store.sendDraft() } },
            enabled = canSend,
            modifier = Modifier.size(44.dp).testTag("respondkit-send"),
            shape = RoundedCornerShape(12.dp),
            contentPadding = PaddingValues(0.dp),
            colors =
                ButtonDefaults.buttonColors(
                    disabledContainerColor = MaterialTheme.colorScheme.primary.copy(alpha = 0.5f),
                    disabledContentColor = MaterialTheme.colorScheme.onPrimary,
                ),
        ) {
            Icon(painterResource(R.drawable.respondkit_send), "Send message", Modifier.size(16.dp))
        }
    }
}

@Composable
private fun Bubble(row: TranscriptRow, isSending: Boolean, retry: (String) -> Unit) {
    BoxWithConstraints(Modifier.fillMaxWidth()) {
        Column(
            Modifier.widthIn(max = maxWidth * 0.84f)
                .align(if (row.customer) Alignment.CenterEnd else Alignment.CenterStart),
            horizontalAlignment = if (row.customer) Alignment.End else Alignment.Start,
            verticalArrangement = Arrangement.spacedBy(4.dp),
        ) {
            SelectionContainer {
                Text(
                    row.text,
                    fontSize = 14.sp,
                    lineHeight = 23.sp,
                    color = if (row.failed) WidgetError else WidgetInk,
                    modifier =
                        Modifier.background(
                                if (row.failed) WidgetError.copy(alpha = 0.1f)
                                else if (row.customer) MaterialTheme.colorScheme.primaryContainer
                                else WidgetFill,
                                RoundedCornerShape(
                                    topStart = 16.dp,
                                    topEnd = 16.dp,
                                    bottomEnd = if (row.customer) 4.dp else 16.dp,
                                    bottomStart = if (row.customer) 16.dp else 4.dp,
                                ),
                            )
                            .padding(horizontal = 12.dp, vertical = 10.dp),
                )
            }
            Row(
                Modifier.heightIn(min = 20.dp).padding(horizontal = 4.dp),
                horizontalArrangement = Arrangement.spacedBy(8.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                row.date?.let {
                    Text(
                        it.format(
                            DateTimeFormatter.ofLocalizedTime(FormatStyle.SHORT)
                                .withLocale(Locale.getDefault())
                        ),
                        fontSize = 12.sp,
                        lineHeight = 16.sp,
                        color = WidgetMuted,
                    )
                }
                row.status?.let { Text(it, fontSize = 12.sp, color = WidgetMuted) }
                row.retryId?.let { id ->
                    TextButton(
                        onClick = { retry(id) },
                        enabled = !isSending,
                        contentPadding = PaddingValues(0.dp),
                    ) {
                        Text(
                            "Try again",
                            fontSize = 12.sp,
                            lineHeight = 16.sp,
                            color =
                                if (row.failed) WidgetError else MaterialTheme.colorScheme.primary,
                        )
                    }
                }
            }
        }
    }
}

File: native/swift/Sources/RespondKitUI/RespondKitScreen.swift
#if os(iOS)
  import SwiftUI
  import RespondKitCore

  private enum WidgetStyle {
    static let foreground = Color(white: 0.09)
    static let muted = Color(white: 0.45)
    static let fill = Color(white: 0.97)
    static let border = Color(white: 0.90)
    static let failure = Color(red: 0.86, green: 0.15, blue: 0.15)
  }

  /// One conversation, presented by the host's fullScreenCover. Matches the web widget's surfaces.
  public struct RespondKitScreen: View {
    private let store: RespondKitStore
    private let title: String
    private let accentColor: Color?
    private let accentForegroundColor: Color
    @Environment(\.dismiss) private var dismiss
    @ScaledMetric(relativeTo: .body) private var titleSize = 16
    @ScaledMetric(relativeTo: .body) private var textSize = 14

    /// Omit accentColor to inherit the host tint (or system/app accent).
    /// Set accentForegroundColor to a dark color when using a light accent.
    public init(
      store: RespondKitStore, title: String = "Support", accentColor: Color? = nil,
      accentForegroundColor: Color = .white
    ) {
      self.store = store
      self.title = title
      self.accentColor = accentColor
      self.accentForegroundColor = accentForegroundColor
    }

    public var body: some View {
      Group {
        if let accentColor { screen.tint(accentColor) } else { screen }
      }
      .preferredColorScheme(.light)
    }

    private var screen: some View {
      VStack(spacing: 0) {
        HStack(spacing: 12) {
          VStack(alignment: .leading, spacing: 0) {
            Text(title).font(.system(size: titleSize, weight: .semibold)).lineLimit(1).frame(
              minHeight: 24)
            Text("Ask us anything").font(.system(size: textSize)).foregroundStyle(WidgetStyle.muted)
              .frame(minHeight: 20)
          }
          Spacer(minLength: 0)
          Button {
            dismiss()
          } label: {
            Image(systemName: "xmark").font(.system(size: 16))
              .frame(width: 40, height: 40).contentShape(Rectangle())
          }
          .buttonStyle(.plain).accessibilityLabel("Close support chat")
          .accessibilityIdentifier("respondkit-close")
        }
        .padding(.horizontal, 16).padding(.vertical, 12)
        Rectangle().fill(WidgetStyle.border).frame(height: 1)
        if let error = store.errorMessage {
          HStack(alignment: .top, spacing: 8) {
            Image(systemName: "exclamationmark.circle")
            Text(error).font(.system(size: textSize))
            Spacer(minLength: 0)
            Button("Retry") { Task { await store.refresh() } }
          }
          .padding(12).background(WidgetStyle.fill)
          .accessibilityIdentifier("respondkit-error")
        }
        ConversationView(store: store, accentForegroundColor: accentForegroundColor)
      }
      .foregroundStyle(WidgetStyle.foreground)
      .background(Color.white.ignoresSafeArea())
      .task { await store.openConversation() }
      .onDisappear { store.setScreenVisible(false) }
    }
  }

  private struct ConversationView: View {
    let store: RespondKitStore
    let accentForegroundColor: Color
    @State private var atBottom = true
    @State private var unseen = 0
    @ScaledMetric(relativeTo: .body) private var textSize = 14

    var body: some View {
      let rows = TranscriptRow.rows(store)
      VStack(spacing: 0) {
        GeometryReader { geometry in
          ScrollViewReader { reader in
            ScrollView {
              LazyVStack(spacing: 12) {
                if rows.isEmpty {
                  if store.isLoading && !store.isSending {
                    VStack(spacing: 16) {
                      ForEach(0..<3) { index in
                        RoundedRectangle(cornerRadius: 10).fill(WidgetStyle.fill)
                          .frame(
                            width: max(0, geometry.size.width - 32) * [0.75, 0.8, 0.67][index],
                            height: [56.0, 80.0, 48.0][index]
                          )
                          .frame(maxWidth: .infinity, alignment: index == 1 ? .leading : .trailing)
                      }
                    }.accessibilityLabel("Loading messages")
                  } else {
                    VStack(spacing: 4) {
                      Text("How can we help?").fontWeight(.medium)
                      Text("Send a message and keep this page open for a quick reply.")
                        .foregroundStyle(WidgetStyle.muted)
                    }
                    .font(.system(size: textSize)).multilineTextAlignment(.center)
                    .frame(maxWidth: .infinity)
                  }
                }
                ForEach(Array(rows.enumerated()), id: \.element.id) { index, row in
                  if let date = row.date,
                    index == 0
                      || rows[index - 1].date.map({
                        !Calendar.current.isDate($0, inSameDayAs: date)
                      }) != false
                  {
                    Text(date.formatted(date: .abbreviated, time: .omitted))
                      .font(.caption).foregroundStyle(WidgetStyle.muted).padding(.vertical, 8)
                  }
                  MessageBubble(
                    row: row, maxWidth: max(0, (geometry.size.width - 32) * 0.84),
                    isSending: store.isSending
                  ) { id in Task { await store.retry(id) } }
                }
                if !rows.isEmpty {
                  Color.clear.frame(height: 2)
                    .onScrollVisibilityChange(threshold: 1) { visible in
                      atBottom = visible
                      if visible { unseen = 0 }
                      guard visible, let id = store.activeThreadID else { return }
                      let cursor = store.loadedCursor
                      Task { await store.markDisplayed(threadID: id, cursor: cursor) }
                    }
                    // Recreate the visibility observer when a new transcript cursor loads.
                    .id("bottom-" + store.loadedCursor)
                }
              }
              .frame(
                maxWidth: .infinity, minHeight: max(0, geometry.size.height - 40),
                alignment: rows.isEmpty
                  ? (store.isLoading && !store.isSending ? .top : .center) : .bottom
              )
              .padding(.horizontal, 16).padding(.vertical, 20)
            }
            .defaultScrollAnchor(.bottom, for: .initialOffset)
            .onChange(of: rows.map(\.id)) { old, new in
              if atBottom || old.isEmpty {
                reader.scrollTo("bottom-" + store.loadedCursor, anchor: .bottom)
              } else {
                unseen += new.filter { !old.contains($0) }.count
              }
            }
            .onChange(of: store.activeThreadID) { _, _ in
              atBottom = true
              unseen = 0
            }
            .overlay(alignment: .bottomTrailing) {
              if unseen > 0 {
                Button {
                  withAnimation { reader.scrollTo("bottom-" + store.loadedCursor, anchor: .bottom) }
                  unseen = 0
                } label: {
                  Label("\(unseen) new", systemImage: "arrow.down")
                    .font(.system(size: textSize)).padding(.horizontal, 12).padding(.vertical, 8)
                    .background(.white, in: Capsule()).overlay(Capsule().stroke(WidgetStyle.border))
                }
                .buttonStyle(.plain).padding(12).accessibilityLabel("Latest messages")
              }
            }
          }
        }
        Rectangle().fill(WidgetStyle.border).frame(height: 1)
        if store.activeThread?.state == "closed" {
          VStack(spacing: 8) {
            Text("This conversation is closed.").foregroundStyle(WidgetStyle.muted)
            Button("Send another message") { store.selectThread(nil) }
          }.font(.system(size: textSize)).padding(12)
        } else {
          Composer(store: store, accentForegroundColor: accentForegroundColor)
        }
      }
    }
  }

  private struct Composer: View {
    let store: RespondKitStore
    let accentForegroundColor: Color
    @FocusState private var focused: Bool
    @ScaledMetric(relativeTo: .body) private var inputSize = 16
    private var canSend: Bool {
      !store.isLoading && !store.isSending
        && !store.draft.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        && store.draft.utf16.count <= 6_000
    }
    var body: some View {
      HStack(alignment: .bottom, spacing: 8) {
        TextField(
          "Write a message…", text: Binding(get: { store.draft }, set: { store.setDraft($0) }),
          axis: .vertical
        )
        .font(.system(size: inputSize)).lineLimit(1...6).focused($focused)
        .padding(.horizontal, 10).padding(.vertical, 10).frame(minHeight: 44)
        .overlay(
          RoundedRectangle(cornerRadius: 12).strokeBorder(
            focused ? AnyShapeStyle(.tint) : AnyShapeStyle(WidgetStyle.border))
        )
        .accessibilityLabel("Message").accessibilityIdentifier("respondkit-composer")
        Button {
          Task { await store.sendDraft() }
        } label: {
          Image(systemName: "paperplane").font(.system(size: 16))
            .foregroundStyle(accentForegroundColor).frame(width: 44, height: 44)
            .background(.tint, in: RoundedRectangle(cornerRadius: 12))
            .opacity(canSend ? 1 : 0.5)
        }
        .buttonStyle(.plain).disabled(!canSend)
        .accessibilityLabel("Send message").accessibilityIdentifier("respondkit-send")
      }.padding(12)
    }
  }

  private struct MessageBubble: View {
    let row: TranscriptRow
    let maxWidth: CGFloat
    let isSending: Bool
    let retry: (String) -> Void
    @ScaledMetric(relativeTo: .body) private var textSize = 14
    @ScaledMetric(relativeTo: .caption) private var captionSize = 12
    var body: some View {
      VStack(alignment: row.customer ? .trailing : .leading, spacing: 4) {
        Text(linkedMessage(row.text)).font(.system(size: textSize)).lineSpacing(5).textSelection(.enabled)
          .foregroundStyle(row.failed ? WidgetStyle.failure : WidgetStyle.foreground)
          .padding(.horizontal, 12).padding(.vertical, 10)
          .background(
            row.failed
              ? AnyShapeStyle(WidgetStyle.failure.opacity(0.1))
              : row.customer
                ? AnyShapeStyle(.tint.opacity(0.1)) : AnyShapeStyle(WidgetStyle.fill),
            in: UnevenRoundedRectangle(
              topLeadingRadius: 16, bottomLeadingRadius: row.customer ? 16 : 4,
              bottomTrailingRadius: row.customer ? 4 : 16, topTrailingRadius: 16))
        HStack(spacing: 8) {
          if let date = row.date { Text(date.formatted(date: .omitted, time: .shortened)) }
          if let status = row.status { Text(status) }
          if let id = row.retryID {
            Button {
              retry(id)
            } label: {
              Label("Try again", systemImage: "exclamationmark.circle")
            }
            .buttonStyle(.plain).foregroundStyle(
              row.failed ? AnyShapeStyle(WidgetStyle.failure) : AnyShapeStyle(.tint)
            )
            .disabled(isSending)
          }
        }.font(.system(size: captionSize)).frame(minHeight: 20).foregroundStyle(WidgetStyle.muted)
          .padding(
            .horizontal, 4)
      }
      .frame(maxWidth: maxWidth, alignment: row.customer ? .trailing : .leading)
      .frame(maxWidth: .infinity, alignment: row.customer ? .trailing : .leading)
    }
  }
#endif
